lib/compat/wordpress-7.0/blocks.php

```diff
@@ -0,0 +1,122 @@
+<?php
+/**
+ * Compatibility shims for block APIs for WordPress 7.0.
+ *
+ * @package gutenberg
+ */
+
+if ( ! function_exists( 'gutenberg_resolve_pattern_blocks' ) ) {
+	/**
+	 * Replaces patterns in a block tree with their content.
+	 *
+	 * @since 6.6.0
+	 * @since 7.0.0 Adds metadata to attributes of single-pattern container blocks.
+	 *
+	 * @param array $blocks An array blocks.
+	 *
+	 * @return array An array of blocks with patterns replaced by their content.
+	 */
+	function gutenberg_resolve_pattern_blocks( $blocks ) {
+		static $inner_content;
+		// Keep track of seen references to avoid infinite loops.
+		static $seen_refs = array();
+		$i                = 0;
+		while ( $i < count( $blocks ) ) {
+			if ( 'core/pattern' === $blocks[ $i ]['blockName'] ) {
+				$attrs = $blocks[ $i ]['attrs'];
+
+				if ( empty( $attrs['slug'] ) ) {
+					++$i;
+					continue;
+				}
+
+				$slug = $attrs['slug'];
+
+				if ( isset( $seen_refs[ $slug ] ) ) {
+					// Skip recursive patterns.
+					array_splice( $blocks, $i, 1 );
+					continue;
+				}
+
+				$registry = WP_Block_Patterns_Registry::get_instance();
+				$pattern  = $registry->get_registered( $slug );
+
+				// Skip unknown patterns.
+				if ( ! $pattern ) {
+					++$i;
+					continue;
+				}
+				//////////////////////////////
+				// START CORE MODIFICATIONS //
+				//////////////////////////////
+				$blocks_to_insert = parse_blocks( trim( $pattern['content'] ) );
+				/*
+				* For single-root patterns, add the pattern name to make this a pattern instance in the editor.
+				* If the pattern has metadata, merge it with the existing metadata.
+				*/
+				if ( count( $blocks_to_insert ) === 1 ) {
+					$block_metadata = $blocks_to_insert[0]['attrs']['metadata'] ?? array();
+					$metadata       = array( 'patternName' => $slug );
+
+					/*
+					* Merge pattern metadata with existing block metadata.
+					* Pattern metadata takes precedence, but existing block metadata
+					* is preserved as a fallback when the pattern doesn't define that field.
+					* Only the defined fields (name, description, categories) are merged;
+					* other metadata keys are not preserved.
```

Comment:
Yeah, just to expand slightly on the nitpicky thread, is it intentional that we're removing other metadata keys? Would this feel like a bug to future metadata-based features that might want to be included within a pattern?

My comment here is very nitpicky because I don't think we have any features as yet that this would impact (i.e. even things like the [allowedBlocks support](https://github.com/WordPress/gutenberg/pull/71986) aren't in `metadata`), so again, feel free to go with what you've got if this is the preferred behaviour!

Reply:
Yep, latest changes address it. Thanks for following up!